tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,227 @@
+
+    *blocked = S2N_BLOCKED_ON_READ;
+    *bytes_read = recvmsg(sock, msg, 0);
+
+    /* The return value will be 0 when the peer has performed an orderly shutdown. */
+    if (*bytes_read == 0) {
+        *bytes_read = 0;
+        s2n_atomic_flag_set(&conn->read_closed);
+        RESULT_BAIL(S2N_ERR_CLOSED);
+    } else if (*bytes_read < 0) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_parse_ancillary_data(struct msghdr *msg, uint8_t *record_type)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(record_type);
+
+#if !S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    RESULT_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+#else
+    RESULT_ENSURE_REF(msg->msg_control);
+    RESULT_ENSURE_GTE(msg->msg_controllen, CMSG_SPACE(sizeof(uint8_t)));
+
+    /* attempt to read the ancillary data */
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    RESULT_ENSURE(hdr != NULL, S2N_ERR_IO);
+
+    /* iterate over all headers until it matches RECORD_TYPE.
+     * CMSG_NXTHDR will return NULL if there are no more cmsg */
+    while (hdr != NULL) {
+        if (hdr->cmsg_level == S2N_SOL_TLS && hdr->cmsg_type == S2N_TLS_GET_RECORD_TYPE) {
+            *record_type = *(unsigned char *) CMSG_DATA(hdr);
+            return S2N_RESULT_OK;
+        }
+
+        /* attempt to get the next header */
+        hdr = CMSG_NXTHDR(msg, hdr);
+    }
+#endif
+
+    /* return an IO error if no record was received */
+    RESULT_BAIL(S2N_ERR_IO);
+}
+
+/* Best practices taken from
+ * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_recv.c.html */
+S2N_RESULT s2n_ktls_recv_msg(struct s2n_connection *conn, int sock, uint8_t *buf, size_t length,
+        uint8_t *record_type, s2n_blocked_status *blocked, ssize_t *bytes_read)
+{
+    RESULT_ENSURE_REF(buf);
+    RESULT_ENSURE_REF(record_type);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(bytes_read);
+    RESULT_ENSURE_GT(length, 0);
+
+    /* Init msghdr */
+    struct msghdr msg = { 0 };
+
+#if S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    /* Allocate a char array of suitable size to hold the ancillary data.
+     * However, since this buffer is in reality a 'struct cmsghdr', use a
+     * union to ensure that it is aligned as required for that structure.
+     */
+    union {
+        /* alloc enough space incase the application recieves more than one cmsg */
+        char buf[CMSG_SPACE(sizeof(uint8_t)) * 4];
+        struct cmsghdr _align;
+    } control_msg = { 0 };
+
+    /* The control message buffer must be zero-initialized in order for the
+     * CMSG_NXTHDR() macro to work correctly.
+     *
+     * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_send.c.html
+     */
+    RESULT_CHECKED_MEMSET(&control_msg.buf, 0, sizeof(control_msg.buf));
```

Comment:
Didn't you already zero-initialize the buffer with `= { 0 }` above?

Reply:
ahh yep, added that after this call so we can remove this. good catch